### DISCUSSION TITLE
Meta issue: much ado about graphics backends

### DISCUSSION THREAD

**Discussion by ctrlcctrlv** [rust-skia/rust-skia#734] [Ideas]
In Skia C++, it is possible to do this:

```c++
void gl_example(int width, int height, void (*draw)(SkCanvas*), const char* path) {
    // You've already created your OpenGL context and bound it.
    sk_sp<const GrGLInterface> interface = nullptr;
    // Leaving interface as null makes Skia extract pointers to OpenGL functions for the current
    // context in a platform-specific way. Alternatively, you may create your own GrGLInterface and
    // initialize it however you like to attach to an alternate OpenGL implementation or intercept
    // Skia's OpenGL calls.
    sk_sp<GrDirectContext> context = GrDirectContext::MakeGL(interface);
    SkImageInfo info = SkImageInfo:: MakeN32Premul(width, height);
    sk_sp<SkSurface> gpuSurface(
            SkSurface::MakeRenderTarget(context.get(), SkBudgeted::kNo, info));
    ...
}
```

Indeed, the [Skia C++ Hello World ](https://github.com/google/skia/blob/main/example/HelloWorld.cpp)trivially switches between Dawn (WebGPU), OpenGL, and Vulkan:

https://github.com/google/skia/blob/d5ec464e9e9c3a295f1fb5e76d1d0e730ebd5019/example/HelloWorld.cpp#L66-L78
```c++
    if (Window::kRaster_BackendType == fBackendType) {
        title.append("Raster");
    } else {
#if defined(SK_GL)
        title.append("GL");
#elif defined(SK_VULKAN)
        title.append("Vulkan");
#elif defined(SK_DAWN)
        title.append("Dawn");
#else
        title.append("Unknown GPU backend");
#endif
    }
```

It certainly is not lost on me that this is only so simple because it's using `skui`, and a real implementation would have to contend with (most likely, as it's ahead of winit on several metrics) SDL2.

But even so, the current fragmented way one must use `rust-skia` to launch graphics backends is highly problematic and leads to a lot of duplicated work.

Why doesn't a simpler way to launch graphics backends through an abstraction over all of them exist in `rust-skia`?

cc @lucasmerlin

**Comment by pragmatrix** [comment]
> Why doesn't a simpler way to launch graphics backends through an abstraction over all of them exist in rust-skia?

Because no one took the time to convert their project specific platform backends so that they all can be used under one single API surface.

And of course I would like that to exist. But I am afraid that this is out of scope of this repository,  though it should be hosted under the rust-skia organization.

It's _technically_ not a big problem to create suitable backends to skia-safe (except perhaps D3D, @tronical built one, and I have one too for winit, though unreleased), but so far everything I saw, and the ones I built (one also vukano based, see: https://github.com/pragmatrix/emergent) are project specific, because - and I would assume - it is _conceptually_ a lot harder to bring them under a single API surface _and_ support the most popular the graphics API wrappers, because I am not sure if they can be mixed.

A plan:

- Collect a list of all skia-safe backends that exist (start with Skulpin, Slint, etc.)
- Create a new project (I am happy to host this under the rust-skia organization)
- Motivate the original maintainers to contribute or just put a lot of hard work into it. 
- Decide what frontend APIs (winit, SDL, probably both) and graphics API backends a MVP will support (vulkano, ash, glutin, whatever).

And so on,

**OR**

Support the Dawn backend in rust-skia and use only WebGPU and hope they are support out all the ways the platforms achieve low latency, window integration, and all the performance tweaks and further unknown unknowns that you can only get aware of when you are deep into that \*hi\*.

Converting to a discussion.
